subprojects/base-services/src/main/java/org/gradle/internal/reflect/DirectInstantiator.java

```diff
@@ -25,6 +25,23 @@
 import java.util.Arrays;
 
 public class DirectInstantiator implements Instantiator {
+    /*
+     * Every time we look up a constructor, we need to pass its types as a class array. Reusing the most common
+     * argument list sizes avoids a lot of garbage.
+     */
+    private static final ThreadLocal<Class[][]> REUSABLE_CLASS_ARRAYS = new ThreadLocal<Class[][]>() {
```

Comment:
It did remove about 10MB of garbage class arrays when configuring the gradle/gradle build.  It doesn't greatly reduce configuration time by itself (no single change in here does), but it all adds up. 

The finally-block takes care of not leaking the Class objects.

The class instances are not really short-lived, they are mostly used for at least the lifetime of a build, often longer (due to classloader caching).

In general I agree that this is the most contentious change in here. I just went through the profiling results and removed one alloc hotspot after another and this was the highest one at the time.